When spawning a proc with a oildrum model, had lua error:

``` lua
[ERROR] addons/starfall/lua/weapons/gmod_tool/stools/starfall_processor.lua:146: Tried to use a NULL entity!
  1. GetClass - [C]:-1
   2. v - addons/starfall/lua/weapons/gmod_tool/stools/starfall_processor.lua:146
    3. unknown - lua/includes/modules/hook.lua:84
```

This is because the entity isn't yet fully initialised when the toolgun hovers over it. Make sure the entity is valid before calling any ent methods on the trace entity. This will also happen for screen as well.

I also see that 'updating' an existing chip doesn't change the model, I'm guessing this is to prevent the previously mentioned physics issues.

--- a/lua/starfall/editor.lua
+++ b/lua/starfall/editor.lua
@@ -659,6 +659,19 @@ if CLIENT then
 			end
 			tabs.selectedTab = SF.Editor.getTabHolder():getTabIndex( SF.Editor.getActiveTab() )
 			file.Write( "sf_tabs.txt", util.TableToJSON( tabs ) )
+
+			local activeWep = LocalPlayer():GetActiveWeapon()
+			if IsValid( activeWep ) and activeWep:GetClass() == "gmod_tool" and activeWep.Mode == "starfall_processor" then
+				local model = nil
+				local ppdata = {}
+				SF.Preprocessor.ParseDirectives( "file", SF.Editor.getCode(), {}, ppdata )
+				if ppdata.models and ppdata.models.file ~= "" then
+					model = ppdata.models.file 
+				end
+
+				local tool = activeWep:GetToolObject( "starfall_processor" )
+				tool.ClientConVar[ "HologramModel" ] = model
+			end 
 		end
 
 		function editor:OnThink ()

--- a/lua/starfall/preprocessor.lua
+++ b/lua/starfall/preprocessor.lua
@@ -156,6 +156,12 @@ local function directive_sharedscreen( args, filename, data )
 end
 SF.Preprocessor.SetGlobalDirective( "sharedscreen", directive_sharedscreen )
 
+local function directive_model( args, filename, data )
+	if not data.models then data.models = {} end
+	data.models[ filename ] = args
+end
+SF.Preprocessor.SetGlobalDirective( "model", directive_model )
+
 --- Mark a file to be included in the upload.
 -- This is required to use the file in require() and dofile()
 -- @name include
@@ -189,3 +195,12 @@ SF.Preprocessor.SetGlobalDirective( "sharedscreen", directive_sharedscreen )
 -- else
 -- \	-- Display result of important calculations
 -- end
+
+--- Set the model of the processor entity.
+-- This does not set the model of the screen entity
+-- @name model
+-- @class directive
+-- @param model String of the model
+-- @usage
+-- \--@model models/props_junk/watermelon01.mdl
+-- -- CODE

--- a/lua/weapons/gmod_tool/stools/starfall_processor.lua
+++ b/lua/weapons/gmod_tool/stools/starfall_processor.lua
@@ -15,7 +15,7 @@ cleanup.Register( "starfall_processor" )
 if SERVER then
 	CreateConVar( "sbox_maxstarfall_processor", 10, { FCVAR_REPLICATED, FCVAR_NOTIFY, FCVAR_ARCHIVE } )
 	
-	function MakeSF ( pl, Pos, Ang, model)
+	function MakeSF ( pl, Pos, Ang, model )
 		if not pl:CheckLimit( "starfall_processor" ) then return false end
 
 		local sf = ents.Create( "starfall_processor" )
@@ -67,24 +67,31 @@ function TOOL:LeftClick ( trace )
 	local Ang = trace.HitNormal:Angle()
 	Ang.pitch = Ang.pitch + 90
 
-	local sf = MakeSF( ply, trace.HitPos, Ang, model )
-
-	local min = sf:OBBMins()
-	sf:SetPos( trace.HitPos - trace.HitNormal * min.z )
-
-	local const = WireLib.Weld( sf, trace.Entity, trace.PhysicsBone, true )
-
-	undo.Create( "Starfall Processor" )
-		undo.AddEntity( sf )
-		undo.AddEntity( const )
-		undo.SetPlayer( ply )
-	undo.Finish()
-
-	ply:AddCleanup( "starfall_processor", sf )
-	
 	if not SF.RequestCode( ply, function ( mainfile, files )
 		if not mainfile then return end
-		if not IsValid( sf ) then return end -- Probably removed during transfer
+
+		local ppdata = {}
+
+		SF.Preprocessor.ParseDirectives( mainfile, files[ mainfile ], {}, ppdata )
+		if ppdata.models and ppdata.models[ mainfile ] and ppdata.models[ mainfile ] ~= "" then
+			model = ppdata.models[ mainfile ]
+		end
+
+		local sf = MakeSF( ply, trace.HitPos, Ang, model )
+
+		local min = sf:OBBMins()
+		sf:SetPos( trace.HitPos - trace.HitNormal * min.z )
+
+		local const = WireLib.Weld( sf, trace.Entity, trace.PhysicsBone, true, true )
+
+		undo.Create( "Starfall Processor" )
+			undo.AddEntity( sf )
+			undo.AddEntity( const )
+			undo.SetPlayer( ply )
+		undo.Finish()
+
+		ply:AddCleanup( "starfall_processor", sf )
+		
 		sf:Compile( files, mainfile )
 	end ) then
 		SF.AddNotify( ply, "Cannot upload SF code, please wait for the current upload to finish.", NOTIFY_ERROR, 7, NOTIFYSOUND_ERROR1 )
@@ -107,6 +114,7 @@ function TOOL:DrawHUD () end
 function TOOL:Think () end
 
 if CLIENT then
+
 	local function get_active_tool ( ply, tool )
 		-- find toolgun
 		local activeWep = ply:GetActiveWeapon()
@@ -114,6 +122,39 @@ if CLIENT then
 
 		return activeWep:GetToolObject( tool )
 	end
+
+	local modelHologram = nil
+
+	hook.Add( "Think", "SF_Update_modelHologram_Processor", function () 
+		if modelHologram == nil or not modelHologram:IsValid() then
+			modelHologram = ents.CreateClientProp()
+			modelHologram:SetRenderMode( RENDERMODE_TRANSALPHA )
+			modelHologram:SetColor( Color( 255, 255, 255, 170 ) )
+			modelHologram:Spawn()
+		end
+
+		local tool = get_active_tool( LocalPlayer(), "starfall_processor" )
+		if tool then
+			local model = tool.ClientConVar[ "HologramModel" ] or tool:GetClientInfo( "Model" )
+			if model and model ~= "" and modelHologram:GetModel() ~= model then
+				modelHologram:SetModel( model )
+			end
+
+			local min = modelHologram:OBBMins()
+			local trace = LocalPlayer():GetEyeTrace()
+
+			if trace.Hit and not ( IsValid( trace.Entity ) and ( trace.Entity:IsPlayer() or trace.Entity:GetClass() == "starfall_processor" ) ) then
+				modelHologram:SetPos( trace.HitPos - trace.HitNormal * min.z )
+				modelHologram:SetAngles( trace.HitNormal:Angle() + Angle( 90, 0, 0 ) )
+				modelHologram:SetNoDraw( false )
+			else
+				modelHologram:SetNoDraw( true )
+			end
+		else
+			modelHologram:SetNoDraw( true )
+		end
+
+	end )
 	
 	hook.Add( "PlayerBindPress", "wire_adv", function ( ply, bind, pressed )
 		if not pressed then return end

--- a/lua/weapons/gmod_tool/stools/starfall_screen.lua
+++ b/lua/weapons/gmod_tool/stools/starfall_screen.lua
@@ -149,7 +149,48 @@ end
 
 if CLIENT then
 	local lastclick = CurTime()
+
+	local function get_active_tool ( ply, tool )
+		-- find toolgun
+		local activeWep = ply:GetActiveWeapon()
+		if not IsValid( activeWep ) or activeWep:GetClass() ~= "gmod_tool" or activeWep.Mode ~= tool then return end
+
+		return activeWep:GetToolObject( tool )
+	end
 	
+	local modelHologram = nil
+
+	hook.Add( "Think", "SF_Update_modelHologram_Screen", function () 
+		if modelHologram == nil or not modelHologram:IsValid() then
+			modelHologram = ents.CreateClientProp()
+			modelHologram:SetRenderMode( RENDERMODE_TRANSALPHA )
+			modelHologram:SetColor( Color( 255, 255, 255, 170 ) )
+			modelHologram:Spawn()
+		end
+
+		local tool = get_active_tool( LocalPlayer(), "starfall_screen" )
+		if tool then
+			local model = tool:GetClientInfo( "Model" )
+			if model and model ~= "" and modelHologram:GetModel() ~= model then
+				modelHologram:SetModel( model )
+			end
+
+			local min = modelHologram:OBBMins()
+			local trace = LocalPlayer():GetEyeTrace()
+
+			if trace.Hit and not ( IsValid( trace.Entity ) and ( trace.Entity:IsPlayer() or trace.Entity:GetClass() == "starfall_screen" ) ) then
+				modelHologram:SetPos( trace.HitPos - trace.HitNormal * min.z )
+				modelHologram:SetAngles( trace.HitNormal:Angle() + Angle( 90, 0, 0 ) )
+				modelHologram:SetNoDraw( false )
+			else
+				modelHologram:SetNoDraw( true )
+			end
+		else
+			modelHologram:SetNoDraw( true )
+		end
+
+	end )
+
 	local function GotoDocs ( button )
 		gui.OpenURL( "http://sf.inp.io" ) -- old one: http://colonelthirtytwo.net/sfdoc/
 	end

